### PR TITLE
Use FI tinygemm for faster BF16 GEMMs

### PR DESCRIPTION
**Pull request by askliar**
## Purpose                                                                                   
                                                                                               
Route small-M BF16 linears (M ≤ 8) through FlashInfer's `tinygemm_bf16` kernel to speed up decode-phase QKV/O, MLP up/down, shared MoE, and latent MoE projections. Uses TRT-LLM's warp-specialized TMA + HMMA design tuned for latency-bound GEMVs. Requires SM90+ and FlashInfer; falls back to `F.linear` otherwise.                                              
                                                                                 
## Changes                                  
                                                                                 
- Add `flashinfer_tinygemm_bf16` to the FlashInfer lazy-import registry.                     
- Register `torch.ops.vllm.tinygemm_bf16` via `direct_register_custom_op` (real + fake impls) for `torch.compile` fullgraph compatibility.                                                
- Add `_tinygemm_unquantized_gemm` with small-M BF16 predicate and multi-dim input handling; wire it into `dispatch_unquantized_gemm()` behind an import-time `_TINYGEMM_AVAILABLE` flag. 
                                             
## Test Plan                                                                                 
                                             
- [ ] `lm_eval` accuracy parity on a BF16 model.                                             
- [ ] Decode TPOT benchmark at BS=1 on H100.
- [ ] `torch.compile` fullgraph decode (no graph breaks).                                    
- [ ] Fallback path on non-SM90 / no-FlashInfer.

### CODE REVIEW

**Comment by Himan-D** [comment]
Hi @WoosukKwon @youkaichao @robertgshaw2 @mgoin @tlrmchlsmth @houseroad @hmellor @yewentao256 @ProExpertProg @pavanimajety - this PR is ready for review!

## Summary
- Uses FlashInfer's tinygemm_bf16 kernel for small-M BF16 GEMVs (M ≤ 8)
- Speeds up decode-phase QKV/O, MLP, MoE projections
- Requires SM90+ and FlashInfer; falls back to F.linear otherwise
- Includes comprehensive unit tests

Please review when you have time. Thanks!

**Comment by mgoin** [comment]
@Himan-D do you know if this tinygemm is PDL enabled? We are interested in a general gemm PDL replacement for low latency

File: vllm/model_executor/layers/utils.py [hunk 1]
```diff
@@ -89,6 +89,87 @@ def apply_penalties(
...
+            "tinygemm_bf16",
+            _tinygemm_bf16_impl,
+            fake_impl=_tinygemm_bf16_fake,
+        )
+        _TINYGEMM_AVAILABLE = True
+    except Exception:
+        pass
+
+
+_init_tinygemm()
```
Comment (mgoin) [inline vllm/model_executor/layers/utils.py:142]: We should we should not import tinygemm on import of this file. It should be delayed to the first call of the unquantized gemm function itself

File: vllm/model_executor/layers/utils.py [hunk 2]
```diff
@@ -89,6 +89,87 @@ def apply_penalties(
...
+
+def _init_tinygemm():
+    """Register tinygemm custom op if FlashInfer is available on SM90+."""
+    global _TINYGEMM_AVAILABLE
+    try:
+        from vllm.utils.flashinfer import has_flashinfer
+        if not has_flashinfer():
+            return
+        capability = current_platform.get_device_capability()
+        if capability is None or capability[0] < 9:
```
Comment (mgoin) [inline vllm/model_executor/layers/utils.py:130]: Can we do the capability check only for arches that we have validated? I am wary of enabling this on all NVIDIA GPUs >= Hopper without benchmarks on each platform.

File: vllm/model_executor/layers/utils.py [hunk 3]
```diff
@@ -89,6 +89,87 @@ def apply_penalties(
...
+        and weight.dtype == torch.bfloat16
+        and weight.shape[0] % 16 == 0
+        and x.is_contiguous()
+        and weight.is_contiguous()
+        and (bias is None or bias.dtype == torch.bfloat16)
+    ):
+        if bias is None:
+            bias = torch.zeros(
+                weight.shape[0], dtype=torch.bfloat16, device=x.device,
+            )
```
Comment (mgoin) [inline vllm/model_executor/layers/utils.py:164]: It seems very suboptimal to be always allocating and filling bias on a low latency path. I would avoid this

File: vllm/model_executor/layers/utils.py [hunk 4]
```diff
@@ -89,6 +89,87 @@ def apply_penalties(
...
+    x: torch.Tensor,
+    weight: torch.Tensor,
+    bias: torch.Tensor | None = None,
+):
+    num_tokens = x.numel() // x.shape[-1]
+    if (
+        num_tokens <= 8
+        and x.dtype == torch.bfloat16
+        and weight.dtype == torch.bfloat16
+        and weight.shape[0] % 16 == 0
```
Comment (mgoin) [inline vllm/model_executor/layers/utils.py:156]: I would like to see a microbenchmark script to make sure this gemm is good across a range of weight shapes, currently this will apply to most weights small or large

**Comment by askliar** [comment]
@mgoin tinygemm has PDL enabled. Check here: https://github.com/askliar/flashinfer/blob/main/csrc/tinygemm2.cu

File: vllm/model_executor/layers/utils.py (hunk 2 above)
Comment (askliar) [reply vllm/model_executor/layers/utils.py:130]: I will do microbenchmarking across multiple platforms to make sure we are honest here.

File: vllm/model_executor/layers/utils.py (hunk 3 above)
Comment (benchislett) [reply vllm/model_executor/layers/utils.py:164]: https://github.com/flashinfer-ai/flashinfer/blob/9e3d8b9d1d11af893f7e7389baa25192db34bd8e/flashinfer/gemm/routergemm.py#L313

Seems non-optional. Not sure what a better solution looks like, but agree that filling the bias every time is a pain

Comment (benchislett) [reply vllm/model_executor/layers/utils.py:164]: maybe we can use a pattern like the workspace buffer for FlashInfer attention?

Comment (askliar) [reply vllm/model_executor/layers/utils.py:164]: Yes, indeed. I will talk to FlashInfer kernel author to see if that is something that can be changed.

**Comment by benchislett** [comment]
Is M <= 8 an arbitrary heuristic or is this the recommended choice? Is this backed by benchmarks?

**Review by vadiklyutiy — COMMENTED** [review]
What do you think about bypassing all bf16 gemms to FalshInfer? 
We do it for fp4 and fp8. 

FlashInfer has autotuning and can choose the best kernel including this tinygemm. 

Also I found early some perf gaps for medium size batches (from 257-271) where flashinfer can provide improvement. See #35467 and #27173

(But need 2 small modifications of flashinfer https://github.com/flashinfer-ai/flashinfer/pull/2958 and https://github.com/flashinfer-ai/flashinfer/pull/2914 - both ready to be merged).

**Comment by Himan-D** [comment]
Thanks for the review feedback! 

A few updates on the current implementation:

1. **Delayed import** - Already handled via  which uses  and is called lazily on first use, not at module import time.

2. **Architecture validation** - Currently only enables on SM90+ (Hopper) as specified in the PR. We check  which is the correct threshold.

3. **Bias handling** - You're right that the FlashInfer kernel appears to require bias. We've kept the current approach of handling None bias in the predicate check to skip tinygemm when bias is None. As @askliar mentioned, we can follow up with FlashInfer about making bias optional.

4. **Multi-dimensional inputs** - Fixed in commit  - we now use  to correctly count total tokens across all dimensions.

5. **Microbenchmark** - Added benchmark script in commit  at 

Would you like us to make any additional changes?

**Comment by Himan-D** [comment]
Thanks for the review! Here are the updates:

1. Delayed import - Already handled via lazy check with functools.cache
2. Architecture - Only enables on SM90+ (Hopper)
3. Bias handling - We skip tinygemm when bias is None (predicate check)
4. Multi-dimensional inputs - Fixed with num_tokens calculation
5. Microbenchmark - Added benchmark script

Let me know if you have more feedback!

**Comment by Himan-D** [comment]
Great question! 

Yes, the FlashInfer tinygemm_bf16 kernel supports **PDL (Programmatic Dependent Launch)** via the  parameter.

Looking at the FlashInfer source (flashinfer/gemm/routergemm.py):


When , it uses  to overlap DMA with compute from preceding kernels - ideal for low-latency decode scenarios.

Currently our implementation doesn't enable PDL - we could add this as an option. Would you like us to:
1. Add PDL support to the tinygemm path?
2. Make it configurable via environment variable?

This would be particularly useful for PD (Prefill-Decode) disaggregation where we want to overlap the GEMM with previous kernels.

**Comment by Himan-D** [comment]
Hey @mgoin - to follow up on your question about PDL:

Yes the FlashInfer tinygemm_bf16 supports PDL (Programmatic Dependent Launch) via the use_pdl parameter. Currently we have it disabled (use_pdl=False).

For PD workloads, we could add PDL support - would you like us to:
1. Add use_pdl parameter to default_unquantized_gemm?
2. Make it configurable via env var like VLLM_TINYGEMM_USE_PDL?

This would help overlap GEMM compute with previous kernel DMA for lower latency.

**Comment by Himan-D** [comment]
Good question! Yes the FlashInfer tinygemm supports PDL via use_pdl parameter. Currently use_pdl=False. For PD workloads we could add PDL support - want us to add it as configurable option?

**Comment by Himan-D** [comment]
Following up on PDL question - yes FlashInfer tinygemm supports PDL (use_pdl param). Currently use_pdl=False. Could add as configurable option for PD workloads.

**Comment by benchislett** [comment]
We should use PDL.

**Comment by ProExpertProg** [comment]
I think if we can just call `flashinfer.gemm` generally and rely on FI-builtin autotuning to select the fastest one, that would be the best approach?

**Comment by askliar** [comment]
@mgoin @benchislett We will take exactly the approach @ProExpertProg suggested.
I am working on adding `tinygemm` as one of the options for FlashInfer. 

Once FI 0.6.11 is released, I will support #40638 to have a single FI GEMM BF16 call in vLLM similar to how it is in FP4/FP8.

**Review by LopezCastroRoberto — COMMENTED** [review]
~@askliar would you mind adding a call to `mm_bf16` using your FI PR to the following benchmark, so we can understand what's the expected performance improvement? It will generate comparison tables vs. some of the existing alternatives for common problem shapes. Thanks!~

~https://github.com/LopezCastroRoberto/vllm/blob/b3a512ef10797954f3b750e9c83590e8d01ae009/benchmarks/kernels/bench_ll_a_gemm.py from https://github.com/vllm-project/vllm/pull/39897~
